addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,448 @@
+		// the cluster is a management cluster
+		clusterRole = clusterRoleManagement
+		clusterClient = r.Client
+	} else {
+		// the cluster is a remote workload cluster
+		clusterRole = clusterRoleWorkload
+		remoteClient, err := r.Tracker.GetClient(ctx, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			log.Error(err, "error getting remote cluster's client")
+			return ctrl.Result{}, err
+		}
+		clusterClient = remoteClient
+		log.Info("successfully got remoteClient")
+		// set watch if not already set. If the watch already exists, it doesn't get re-created
+		if err := r.watchPackageInstalls(ctx, cluster, log); err != nil {
+			log.Error(err, "error watching PackageInstalls on target cluster")
+			return ctrl.Result{}, err
+		}
+		log.Info("finished setting up remote watch for the cluster")
+	}
+
+	if err := r.reconcile(ctx, clusterClient, cluster, clusterRole, log); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcile iterates over all core/additional packages and calls reconcileClusterBootstrapStatus() for each.
+// it eventually updates ClusterBootstrapStatus with the condition entries (one for each package) in a single update operation
+// Note: use of the client c (remote or not) depends on whether the corresponding cluster is workload or management which is
+// determined in the Reconcile function
+func (r *PackageInstallStatusReconciler) reconcile(ctx context.Context, clusterClient client.Client, cluster *clusterapiv1beta1.Cluster, clusterRole ClusterRole, log logr.Logger) error {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	// ClusterBootstrap resource exists in the management cluster, that's why we need to use local client instead of remote client for fetching it
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(cluster), clusterBootstrap); err != nil {
+		return err
+	}
+	clusterBootstrap = clusterBootstrap.DeepCopy()
+
+	// this shouldn't include kapp ctrl package as it'll get processed separately
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+	}, clusterBootstrap.Spec.AdditionalPackages...)
+
+	var cbStatusChanged bool
+
+	for _, pkg := range packages {
+		if pkg == nil {
+			continue
+		}
+		cbConditionChanged, err := r.reconcileClusterBootstrapStatus(ctx, clusterClient, &clusterBootstrap.Status, cluster.Name, clusterRole, pkg.RefName, constants.TKGSystemNS, log)
+		cbStatusChanged = cbStatusChanged || cbConditionChanged
+		if err != nil {
+			// in this case, we delete the existing condition corresponding to the package in the ClusterBootstrapStatus (if existing); as the corresponding pkgi or package resources do not exist for the package anymore
+			// then just continue with collecting PackageInstallStatus for other packages
+			cbStatusChanged = cbStatusChanged || r.removeConditionIfExistsForPkgName(&clusterBootstrap.Status, pkg.RefName)
```

Comment:
Q: doesn't Go short-circuit that `II`? This would mean that `r.removeConditionIfExistsForPkgName` is only executed if `cbStatusChanged` is false

Reply:
Oops, very true. Thanks for catching this!